### ISSUE TITLE
Deadlock in library_free at subprocess exit

### ISSUE THREAD

**Issue by Izzette** [ffi/ffi#942]
Hey, we're experiencing frequent issues with a deadlock occurring when our ruby 3.0.3 subprocesses exits.  It appears to be part of a finaliser ([process.c:4348](https://github.com/ruby/ruby/blob/3fb7d2cadc18472ec107b14234933b017a33c14d/process.c#L4348) → ... → [gc.c:3666](https://github.com/ruby/ruby/blob/3fb7d2cadc18472ec107b14234933b017a33c14d/gc.c#L3666)) for ffi ([DynamicLibrary.c:171](https://github.com/ffi/ffi/blob/master/ext/ffi_c/DynamicLibrary.c#L171)), the last call being made in the main thread before hitting glibc.  Several threads are "exiting" and stuck in `__lll_lock_wait_private` called by `__deallocate_stack`.  The child process does not respond to SIGTERM, and is stuck here for almost 20 minutes.  This occurs about 25 times a day for us.

The main thread is spinning on a futex:
```text
(gdb) display/4i $pc-4
6: x/4i $pc-4
   0x7f619f7c9c3a <__wait_lookup_done+186>:	mov    %ebx,%eax
   0x7f619f7c9c3c <__wait_lookup_done+188>:	syscall
=> 0x7f619f7c9c3e <__wait_lookup_done+190>:	cmp    $0xfffffffffffff000,%rax
   0x7f619f7c9c44 <__wait_lookup_done+196>:	jbe    0x7f619f7c9c20 <__wait_lookup_done+160>
 ```
 
 Full trace is:
 ```text
 (gdb) bt
#0  0x00007f619f7c9c3e in __wait_lookup_done () from /lib/x86_64-linux-gnu/libpthread.so.0
#1  0x00007f619fde96ea in ?? () from /lib64/ld-linux-x86-64.so.2
#2  0x00007f619fde986e in ?? () from /lib64/ld-linux-x86-64.so.2
#3  0x00007f619f225b2f in _dl_catch_exception () from /lib/x86_64-linux-gnu/libc.so.6
#4  0x00007f619f225bbf in _dl_catch_error () from /lib/x86_64-linux-gnu/libc.so.6
#5  0x00007f619f470975 in ?? () from /lib/x86_64-linux-gnu/libdl.so.2
#6  0x00007f619f470364 in dlclose () from /lib/x86_64-linux-gnu/libdl.so.2
#7  0x00007f618d6f7c31 in library_free (library=0x7f6188dd03c0) at DynamicLibrary.c:171
#8  0x00007f619faed656 in run_final (zombie=140056861321360, objspace=0x7f619ea5c000) at gc.c:3666
#9  finalize_list (objspace=objspace@entry=0x7f619ea5c000, zombie=140056861321360) at gc.c:3685
#10 0x00007f619faf898d in rb_objspace_call_finalizer (objspace=0x7f619ea5c000) at gc.c:3848
#11 0x00007f619fadaa81 in rb_ec_finalize (ec=0x7f619ea1e050) at vm_core.h:1789
#12 rb_ec_cleanup (ec=<optimized out>, ex=<optimized out>) at eval.c:295
#13 0x00007f619fadacfc in ruby_cleanup (ex=0) at vm_core.h:1809
#14 0x00007f619fadad09 in ruby_stop (ex=<optimized out>) at eval.c:329
#15 0x00007f619fbbdb18 in rb_f_fork (obj=<optimized out>) at process.c:4348
#16 rb_f_fork (obj=<optimized out>) at process.c:4338
#17 0x00007f619fc8418b in vm_call_cfunc_with_frame (ec=0x7f619ea1e050, reg_cfp=0x7f619eb63478, calling=<optimized out>) at vm_insnhelper.c:2928
#18 0x00007f619fc9271a in vm_sendish (method_explorer=<optimized out>, block_handler=<optimized out>, cd=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>) at vm_callinfo.h:336
#19 vm_exec_core (ec=0x7f619ea1e050, initial=128) at insns.def:876
#20 0x00007f619fc95586 in rb_vm_exec (ec=0x7f619ea1e050, mjit_enable_p=true) at vm.c:2163
#21 0x00007f619fc9ec55 in invoke_block (captured=<optimized out>, opt_pc=<optimized out>, type=<optimized out>, cref=0x0, self=140057131730720, iseq=0x7f61985f3d30, ec=0x7f619ea1e050) at vm.c:1263
#22 invoke_iseq_block_from_c (me=0x0, is_lambda=<optimized out>, cref=0x0, passed_block_handler=0, kw_splat=0, argv=0x0, argc=0, self=140057131730720, captured=<optimized out>, ec=0x7f619ea1e050)
    at vm.c:1335
#23 invoke_block_from_c_bh (argc=<optimized out>, argv=<optimized out>, kw_splat=<optimized out>, passed_block_handler=<optimized out>, cref=<optimized out>, is_lambda=<optimized out>,
    force_blockarg=<optimized out>, block_handler=<optimized out>, ec=<optimized out>) at vm.c:1353
#24 vm_yield (argc=0, argv=0x0, kw_splat=0, ec=0x7f619ea1e050) at vm.c:1398
#25 rb_yield_0 (argv=0x0, argc=0) at vm_eval.c:1331
#26 loop_i (_=_@entry=0) at vm_eval.c:1430
#27 0x00007f619fad8413 in rb_vrescue2 (b_proc=0x7f619fc9e9f0 <loop_i>, data1=0, r_proc=0x7f619fc7cf00 <loop_stop>, data2=0, args=args@entry=0x7fff0d328cf0) at eval.c:1019
#28 0x00007f619fad868a in rb_rescue2 (b_proc=<optimized out>, data1=<optimized out>, r_proc=<optimized out>, data2=<optimized out>) at eval.c:996
#29 0x00007f619fc8418b in vm_call_cfunc_with_frame (ec=0x7f619ea1e050, reg_cfp=0x7f619eb636a8, calling=<optimized out>) at vm_insnhelper.c:2928
#30 0x00007f619fc9cdd9 in vm_call_method_each_type (ec=0x7f619ea1e050, cfp=0x7f619eb636a8, calling=0x7fff0d328fb0) at vm_insnhelper.c:3418
#31 0x00007f619fc9d634 in vm_call_method (ec=0x7f619ea1e050, cfp=0x7f619eb636a8, calling=<optimized out>) at vm_insnhelper.c:3536
#32 0x00007f619fc8811c in vm_sendish (ec=0x7f619ea1e050, reg_cfp=0x7f619eb636a8, cd=0x7f6197c413e0, block_handler=<optimized out>, method_explorer=mexp_search_method) at vm_callinfo.h:336
#33 0x00007f619fc8fe78 in vm_exec_core (ec=0x7f619ea1e050, initial=128) at insns.def:770
#34 0x00007f619fc95d5d in rb_vm_exec (ec=0x7f619ea1e050, mjit_enable_p=true) at vm.c:2172
#35 0x00007f619fad50b9 in rb_ec_exec_node (ec=ec@entry=0x7f619ea1e050, n=n@entry=0x7f619795c978) at eval.c:317
#36 0x00007f619fadad65 in ruby_run_node (n=0x7f619795c978) at eval.c:375
#37 0x00005615b0ee610b in main (argc=<optimized out>, argv=<optimized out>) at ./main.c:50
```
Frame 15 (`rb_f_fork`) is the last frame from the parent process, observing the ruby stack trace by unwinding `rb_execution_context_t` also shows that the VM context has reached the fork-point.

I have a coredump, I cannot share it, but if you have ideas about how to collect more data, I'm able to add it here.  I spent a little time trying to find more information about what library is being dlclosed, but I'm not sure how to get more data about it.

**Comment by shanet** [comment]
This is kind of an old issue at this point, but I ran into the same problem. There's a lengthy thread about it (with a solution!) over at https://github.com/DataDog/dd-trace-rb/issues/3015.

tl;dr though: [Per POSIX](https://www.man7.org/linux/man-pages/man2/fork.2.html):

> After a fork() in a multithreaded program, the child can safely call only async-signal-safe functions

Where `dlclose` is not a signal-safe function which is why these deadlocks are happening.

Ideally only single threaded program would be forking which should avoid this problem. However, if it is necessary to fork a multi-threaded program `at_exit {exit!(0)}` can be used to bypass Ruby's calling of FFI's finalizer method which calls `dlclose`. See the linked thread on the dd-trace-rb repo for caveats to this approach through (for example, ensuring that any running background threads flush their buffers first to avoid data loss).
